Lurk/Evaluation/Eval.lean

```diff
@@ -1,260 +1,132 @@
-        return .lam ns' patch lb
-    | v => throw s!"expected lambda value, got\n {v}"
-  | .quote s => return .ofSExpr s
-  | .binaryOp op e₁ e₂ => do 
-    evalBinaryOp (← evalM env e₁ (iter + 1)) (← evalM env e₂ (iter + 1)) op
-  | .atom e => do 
-    return match ← evalM env e (iter + 1) with
-    | .cons .. => TRUE
-    | _ => FALSE
-  | .cons e₁ e₂ => do 
-    return .cons (← evalM env e₁ (iter + 1)) (← evalM env e₂ (iter + 1))
-  | .strcons e₁ e₂ => do 
-    match (← evalM env e₁ (iter + 1)), (← evalM env e₂ (iter + 1)) with
-    | .lit (.char c), .lit (.str s) => return .lit (.str ⟨c :: s.data⟩)
-    | .lit (.char _), v => throw s!"expected string value, got\n {v}"
-    | v, _ => throw s!"expected char value, got\n {v}"
-  | .car e => do 
-    match ← evalM env e (iter + 1) with
-    | .cons v _ => return v
-    | .lit (.str s) => match s.data with
-      | c::_ => return .lit $ .char c
-      | [] => return FALSE
-    | v => throw s!"expected cons value, got\n {v}"
-  | .cdr e => do
-    match ← evalM env e (iter + 1) with
-    | .cons _ v => return v
-    | .lit (.str s) => match s.data with
-      | _::cs => return .lit $ .str ⟨cs⟩
-      | [] => return .lit $ .str ""
-    | v => throw s!"expected cons value, got\n {v}"
-  | .emit e => do
-    let v ← evalM env e (iter + 1)
-    dbg_trace v
-    pure v
-  | .begin e₁ e₂ => do
-    discard $ evalM env e₁ (iter + 1)
-    evalM env e₂ (iter + 2)
-  | .currEnv => do
-    throw "floating `current-env`, try `(current-env)` instead"
-  -- let t_val : Format := ← match env.find? `getelem with 
-  --   | some v => do return (← v).pprint
-  --   | none => return "not there"
-  -- dbg_trace s!"[`t] ....\n{t_val}\n"
-  -- dbg_trace s!"[result of {iter}] ....\n{v.pprint}\n"
-  | .commit e => evalM env (.hide (.mkNum 0) e)
-  | _ => throw "TODO"
-  return v
 end
 
-def eval (e : Expr) : Except String Value :=
-  match (ExceptT.run (evalM default e)) () with
-  | .ok res _ => res
-  | .error _ _ => .error "FIXME"
-
-def ppEval (e : Expr) : IO Format :=
-  match (ExceptT.run (evalM default e)) () with
-  | .ok res _ => match res with
-    | .ok v => pure v.pprint
-    | .error s => throw $ .otherError 0 s -- FIXME
-  | .error _ _ => throw $ .otherError 0 "HERE" -- FIXME
-
-instance : OfNat Value n where 
-  ofNat := .lit $ .num $ Fin.ofNat n
-
-instance : Coe Char Value where 
-  coe c := .lit (.char c)
-
-instance : Coe String Value where 
-  coe s := .lit (.str s)
-
-instance : Coe (List (Name × Nat)) Value where
-  coe l := .env $ l.map fun (name, n) => (name, .lit $ .num $ Fin.ofNat n)
+def Value.num : Value → Except String F
+  | .ast (.num x) => pure $ .ofNat x
+  | v => throw s!"expected numerical value, got\n  {v.toString}"
+
+partial def Expr.evalOp₁ (op : Op₁) (v : Value) : Except String Value := match op, v with
+  | .atom, .ast (.cons ..) => return .ast .nil
+  | .atom, _ => return .ast (.sym "T")
```

Comment:
We have `.lit .t`

Reply:
Sorry I kinda made some mistakes along the way doing this. I reverted back to using `Lit` and stuff, and will push soon